cryptol-remote-api/python/cryptol/cryptoltypes.py

```diff
@@ -229,230 +304,183 @@ def is_plausible_json(val : Any) -> bool:
         return f"({self.left} % {self.right})"
 
-    def __repr__(self) -> str:
-        return f"Mod({self.left!r}, {self.right!r})"
-
-class CeilMod(CryptolType):
-    def __init__(self, left : CryptolType, right : CryptolType) -> None:
-        self.left = left
-        self.right = right
-
+@dataclass
+class CeilMod(BinaryType):
     def __str__(self) -> str:
         return f"({self.left} %^ {self.right})"
 
-    def __repr__(self) -> str:
-        return f"CeilMod({self.left!r}, {self.right!r})"
-
-class Expt(CryptolType):
-    def __init__(self, left : CryptolType, right : CryptolType) -> None:
-        self.left = left
-        self.right = right
-
+@dataclass
+class Expt(BinaryType):
     def __str__(self) -> str:
         return f"({self.left} ^^ {self.right})"
 
-    def __repr__(self) -> str:
-        return f"Expt({self.left!r}, {self.right!r})"
-
+@dataclass
 class Log2(CryptolType):
-    def __init__(self, operand : CryptolType) -> None:
-        self.operand = operand
+    operand : CryptolType
 
     def __str__(self) -> str:
         return f"(lg2 {self.operand})"
 
-    def __repr__(self) -> str:
-        return f"Log2({self.operand!r})"
-
+@dataclass
 class Width(CryptolType):
-    def __init__(self, operand : CryptolType) -> None:
-        self.operand = operand
+    operand : CryptolType
 
     def __str__(self) -> str:
         return f"(width {self.operand})"
 
-    def __repr__(self) -> str:
-        return f"Width({self.operand!r})"
-
-class Max(CryptolType):
-    def __init__(self, left : CryptolType, right : CryptolType) -> None:
-        self.left = left
-        self.right = right
-
+@dataclass
+class Max(BinaryType):
     def __str__(self) -> str:
         return f"(max {self.left} {self.right})"
 
-    def __repr__(self) -> str:
-        return f"Max({self.left!r}, {self.right!r})"
-
-class Min(CryptolType):
-    def __init__(self, left : CryptolType, right : CryptolType) -> None:
-        self.left = left
-        self.right = right
-
+@dataclass
+class Min(BinaryType):
     def __str__(self) -> str:
         return f"(min {self.left} {self.right})"
 
-    def __repr__(self) -> str:
-        return f"Min({self.left!r}, {self.right!r})"
+@dataclass
+class LenFromThenTo(CryptolType):
+    num_from : CryptolType
+    num_then : CryptolType
+    num_to : CryptolType
+
+    def __str__(self) -> str:
+        return f"(lengthFromThenTo {self.num_from} {self.num_then} {self.num_to})"
 
+@dataclass
 class Tuple(CryptolType):
-    types : Iterable[CryptolType]
+    types : typing.Sequence[CryptolType]
 
     def __init__(self, *types : CryptolType) -> None:
         self.types = types
 
     def __repr__(self) -> str:
-        return "Tuple(" + ", ".join(map(str, self.types)) + ")"
+        return "Tuple(" + ", ".join(map(repr, self.types)) + ")"
 
-class Record(CryptolType):
-    def __init__(self, fields : Dict[str, CryptolType]) -> None:
-        self.fields = fields
+    def __str__(self) -> str:
```

Comment:
Is `Record` not supposed to have a `__repr__`? It's not obvious to me yet why some classes have `__repr__` and some don't. Ah, is it because `@dataclass` is giving us the one we want for some but not others?

Reply:
Oops, pretend this comment is a few lines down where `Record` is defined xD

Reply:
Added a comment saying why `Tuple` has an `__init__` and `__repr__` while all the others don't:
```python
    # we override the @dataclass __init__ and __repr__ because we want the
    #  syntax of variable numbers of arguments
```
Does that make sense?

Reply:
Yes